### PR TITLE
[WIP] Metric Taxonomy

### PR DESCRIPTION
**Pull request by blue42u**
Initial, very WIP metric taxonomy support. Will eventually resolve https://github.com/HPCToolkit/hpctoolkit/issues/432.

Relative to the rough design proposed there, there are some additions to the [`<Metric>` tags in experiment.xml](https://github.com/HPCToolkit/hpctoolkit/blob/52818b64bae84f055a040a670e8b510e6a94c903/lib/dtd/hpc-experiment.dtd#L24-L45) and to the [input metrics in METRICS.yaml](https://github.com/HPCToolkit/hpctoolkit/blob/52818b64bae84f055a040a670e8b510e6a94c903/doc/METRICS.yaml#L21-L33). In short, measured metrics are now uniquely identified by a tuple of 4 strings:
 - `name`: arbitrary string used to name the exact metric that was measured,
 - `scope`: arbitrary string used to separate propagations (inclusive vs. exclusive, etc.),
 - `formula`: special-format string specifying the calculation applied to raw measurement values prior to combination, and
 - `combine`: string-enumeration specifying the operator to use for combinations.

In terms of the "accumulate" -> "combine" -> "finalize" view of analysis processing, `formula` specifies "accumulate" and `combine` specifies "combine," "finalize" comes from the formulas within the `METRICS.yaml`. All formulas are needed since non-summary data is raw and will need to be processed through these functions to generate a thread-local view with full metric taxonomy (i.e. miniature Prof on a small subset of the data). For this reason `<MetricDB>` tags are folded into the corresponding `<Metric>` tags, using a secondary identifier attribute `raw-i` to specify the id to use for non-summary profiles (to disambiguate the identifier attribute for the summary profile is now `summary-i`).

Currently this branch presumes Prof2 is in charge of "interpreting" the measured Metrics, rather than Run specifying what it wants to happen (in part because there is no clear answer to a "mixed" measurement with differing metrics in different threads). Whether this approach is sufficiently scalable remains to be seen.

### CODE REVIEW

File: src/lib/profile/sinks/metricsyaml.cpp [hunk 1]
```diff
@@ -122,17 +118,469 @@ static YAML::Emitter& operator<<(YAML::Emitter& e, const MetricScope ms) {
...
+        for(const Metric& sm: sub_i) rawLeaf(out, sm);
+        out << EndSeq
+            << EndMap;
+      }
+      out << EndSeq
+          << EndMap;
+    }
+  } c_memcpy;
+
+  struct MemSet {
```
Comment (Jokeren) [inline src/lib/profile/sinks/metricsyaml.cpp:369]: Do those structures need to be hard-coded? Or we can generate code parsing code from a configuration file (i.e., yaml or protocol buffer)?

Comment (blue42u) [reply src/lib/profile/sinks/metricsyaml.cpp:369]: At the moment I haven't worked out how best to dynamically generate the taxonomic structure. I hard-coded the whole thing to get a minimal example together for experimental Viewer support. If you have an idea for an approach please elaborate,  suggestions would be very welcome.

**Comment by aarontcopal2** [comment]
@blue42u , Is it possible to add an additional string for each metric called `format`? With that flexibility, one can represent metrics as floating-point or integer.

**Comment by blue42u** [comment]
@aarontcopal2 `format` already exists in this PR? https://github.com/HPCToolkit/hpctoolkit/pull/481/files#diff-4598fc89d83c459cb8f8955220e821a27211179855ce63e059a89b4b282acd7eR114-R121

For integers I would recommend `%.3g`, that (should as per `printf`) render as integers until 1000 at which point it flips to sci-notation.

**Comment by blue42u** [comment]
Rebased and superseded by #495.

**Pull request closed without merging**